File: model/src/main/java/pl/agh/harmonytools/model/HarmonicFunction.java
package pl.agh.harmonytools.model;

import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.Setter;

import java.util.List;

@Getter
@Setter
@Builder
@AllArgsConstructor
public class HarmonicFunction {

    private String name;
    private int degree;
    private ChordComponent position;
    private ChordComponent revolution;
    private Delay delay;
    private List<ChordComponent> extra;
    private List<ChordComponent> omit;
    private Boolean isDown;
    private ChordSystem system;
    private Mode mode;
    private Key key;

}

File: model/src/main/scala/pl/agh/harmonytools/model/HarmonicFunction.scala
package pl.agh.harmonytools.model

import pl.agh.harmonytools.model.util.ChordComponentManager

case class HarmonicFunction(
  baseFunction: FunctionNames.BaseFunction,
  degree: ScaleDegree.Degree,
  position: Option[ChordComponent],
  revolution: ChordComponent,
  delay: List[Delay],
  extra: List[ChordComponent],
  omit: List[ChordComponent],
  isDown: Boolean,
  system: ChordSystem.System,
  mode: Mode.BaseMode,
  key: Option[Key],
  isRelatedBackwards: Boolean
)

object HarmonicFunction {
  def apply(
    baseFunction: FunctionNames.BaseFunction,
    degree: Option[ScaleDegree.Degree],
    position: Option[ChordComponent] = None,
    revolution: ChordComponent = ChordComponentManager.getRoot,
    delay: List[Delay] = List.empty,
    extra: List[ChordComponent] = List.empty,
    omit: List[ChordComponent] = List.empty,
    isDown: Boolean = false,
    system: ChordSystem.System = ChordSystem.UNDEFINED,
    mode: Mode.BaseMode = Mode.MAJOR,
    key: Option[Key] = None,
    isRelatedBackwards: Boolean = false
  ): HarmonicFunction = {
    degree match {
      case Some(value) => new HarmonicFunction(baseFunction, value, position, revolution, delay, extra, omit, isDown, system, mode, key, isRelatedBackwards)
      case None => new HarmonicFunction(baseFunction, baseFunction.baseDegree, position, revolution, delay, extra, omit, isDown, system, mode, key, isRelatedBackwards)
    }
  }
}
